### ISSUE TITLE
pkgname parsing for string perl-PerlIO-utf8_strict-0.007_1

### ISSUE THREAD

**Issue by svenper** [void-linux/void-packages#14005]
### System

* xuname:  
  `Void 5.2.10_1 x86_64-musl GenuineIntel notuptodate hold rrnFFFFFFFFFFF`
* package:  
  TBA

### Expected behavior

A package named `perl-PerlIO-utf8_strict` works as a dependency.

### Actual behavior

If a package depends on `perl-PerlIO-utf8_strict`, xbps-src will instead look for `perl-PerlIO` which could exist (`PerlIO` module is part of core `perl` but not generated to `provides`).

### Steps to reproduce the behavior

Build #13965

**Comment by svenper** [comment]
Seems like a red herring for the `_` in the pkgname. If i replace the underscore with a dash in the package name and add `wrksrc="${wrksrc/-strict/_strict}"` to the template, the package is treated nicer by xbps-src. (In context of building #13965)

So the string `perl-PerlIO-utf8_strict` looks like it is parsed something like
* pkgname `perl-PerlIO`
* version `utf8`
* revision `strict` (?)

Many other packages are named by the pattern \*-\*_\* so maybe the `8` in `utf8` is causing confusion? Should the package be changed or the parser?

**Comment by svenper** [comment]
```text
$ xbps-uhelper getpkgname perl-PerlIO-utf8_strict-0.007_1
perl-PerlIO-utf8_strict
```

↑ good

```text
$ xbps-uhelper getpkgname perl-PerlIO-utf8_strict
perl-PerlIO
```

↑ weird, should fail?

```text
$ xbps-uhelper getpkgname perl-PerlIO
Invalid string, expected <string>-<version>_<revision>
```

↑ good

**Comment by Chocimier** [comment]
Specifying `perl-PerlIO-utf8_strict>0` help.
Robust solution may be:
- only recognize revision if it is integer, and
- disallow pkgnames that look like pkgver, or end with underscore-integer (simpler, but we have `rtl_433`)

**Comment by svenper** [comment]
Hmm, neither `perl-PerlIO-utf8_strict>0` nor even `perl-PerlIO-utf8_strict-0.007_1` work as dependencies. Looks like [this section in xbps](https://github.com/void-linux/xbps/blob/7b66a2885b38a5af58f2b17c9b3bc9bda6a42ad3/lib/util.c#L224-L256) needs to be changed.

**Comment by svenper** [comment]
Unless xbps switches to regular expressions, I would think it works reasonably well to check that whatever comes after the last `_` is a number. All existing packages already only have numbers in `revision` so the check should affect nothing. I don't understand why this segfaults (tested with `xbps-install -Snu`).

```diff
--- lib/util.c
+++ lib/util.c
@@ -236,7 +236,7 @@ xbps_pkg_name(const char *pkg)
        for (unsigned int i = 0; i < strlen(p); i++) {
                if (p[i] == '_')
                        break;
-               if (isdigit((unsigned char)p[i]) && strchr(p, '_')) {
+               if (isdigit((unsigned char)p[i]) && strchr(p, '_') && isdigit((unsigned char)strrchr(p, '_')[1])) {
                        valid = true;
                        break;
                }

```

**Comment by Chocimier** [comment]
Any new info? I am going to look at this today, as last comment suggest that xbps-src tries to remove version more that once from same pkgver.

**Comment by Chocimier** [comment]
So for hostmakedepends only pkgnames, not pkgpatterns are allowed what makes sense, because it is known what version of package is specified in template. PR for xbps side is sent.

**Comment by svenper** [comment]
Thanks for looking into it! No news from me. I applied your commit to xbps-0.56_7 and the uhelper command work well on my end now and I am able to build #13965 .

**Comment by xtraeme** [comment]
This has been fixed since 0.57.1.
